homeassistant/helpers/device_registry.py

```diff
@@ -0,0 +1,74 @@
+"""Provide a way to connect entities belonging to one device.
+"""
+
+import logging
+
+import attr
+
+from homeassistant.core import callback
+from homeassistant.loader import bind_hass
+
+_LOGGER = logging.getLogger(__name__)
+
+DATA_REGISTRY = 'device_registry'
+
+
+@attr.s(slots=True, frozen=True)
+class DeviceEntry:
+    """Device Registry Entry."""
+
+    unique_id = attr.ib(type=str)
```

Comment:
Let's name this `id` and use a factory to automatically generate it

```python
    id = attr.ib(type=str, default=attr.Factory(lambda: uuid.uuid4().hex))
```

Reply:
So after reading everything, let's circle back here:

We should add the `id` property to `DeviceEntry`, it will be how we can point at this device from the entity registry.

For unique_id and serial, I would expect us to have a tuple of identifiers. Like connection, but with just identifying things:

```
identifiers = (
  ('serial', 'abcdefgh'),  # if we can read the device serial
  ('nest', 'czxcz'),  # if the developer wants to pass an identifier. They can use their domain to scope it.
)
```

Reply:
Then we would only do identification on identifiers, or connections are still viable due to discovery?